DIPs/1NNN-RA.md

```diff
@@ -0,0 +1,409 @@
+```D
+safe unittest
+{
+    class Book : ProtoObject, Orderd
+    {
+        mixin ImplementOrdered!("x");
+        enum BookFormat { pdf, epub, paperback, hardcover }
+
+        ssize_t isbn;
+        BookFormat format;
+
+        this(ssize_t isbn, BookFormat format)
+        {
+            this.isbn = isbn;
+            this.format = format;
+        }
+    }
+    
+    auto b1 = new Book(12345, Book.BookFormat.pdf);
+    auto b2 = new Book(12345, Book.BookFormat.paperback);
+    assert(b1.cmp(b2) == 0);
+}
+```
+ 
+To maximize the flexibility of the user, `ImplementOrdered` also accepts the comparator function to apply on the members.
+```D
+@safe unittest
+{
+    class Widget : ProtoObject, Ordered
+    {
+        mixin ImplementOrdered!("x", "y", (int a, int b) => a - b, "z", (int a, int b) => a - b + 1, "t");
+        int x, y, z, t;
+        /* code */
+    }
+}
+```
+
+There are cases where it is desirable to implement the default comparison operation on almost all the fields of the aggregate. In such cases, especially if there are a lot of member fields, it would be unpleasant to pass `ImplementOrdered` the name of all the fields we want to take into account while comparing; it would be cleaner and easier to inform it about the exceptions. This is why we defined `ImplementOrderedExcept`.
+```D
+@safe unittest
+{
+    class Widget : ProtoObject, Ordered
+    {
+        mixin ImplementOrderedExcept!("unorderableField");
+        int x, y, z, t;
+        size_t unorderableField;
+        /* code */
+    }
+}
+```
+
+#### Equality
+
+Checking for equality follows in the footsteps of ordering, and requires the user to implement the `Equals` interface.
+```D
+interface Equals
+{
+    const @nogc nothrow pure @safe scope
+    int equals(scope const ProtoObject rhs);
+}
+```
+
+Again, as it is the case with ordering, `rhs` is a `ProtoObject` so we can treat the worst-case scenario: compare two objects and all we know is that they are `ProtoObject`s.
+
+```D
+bool __equals(ProtoObject p1, ProtoObject p2)
+{
+    if (p1 is p2) return true;
+    Equals o1 = cast(Equals) p1;
+    Equals o2 = cast(Equals) p2;
+    if (o1 is null || o2 is null) return false;
+    return o1.equals(o2) && o2.equals(o1);
+}
+```
+
+An important note: it isn't sufficient for `o1` to say it's equal to `o2`; `o2` must also agree. Think about comparing a `Base` and a `Derived` instance. The base could be equal to the derived, but the derived might have some extra data so it won't be equal to the base instance.
+
+As it was the case with `Ordered`, to aid the user with the boilerplate code, we provide `ImplementEquals` and `ImplementEqualsExcept` that respect the same design. "Consistency, consistency, consistency" (Scott Meyers).
+
+#### Hashing
+
+The user must implement the `Hash` interface if he desires to provide a hashable behaviour.
+```D
+interface Hash
+{
+    const @nogc nothrow pure @safe scope
+    size_t toHash();
```

Comment:
Since old code won't break unlike how I assumed, yeah I guess `nothrow` is justified. I still do question `@nogc` though. An ideal hash function can work in stack only, but it may sometimes be easier to use the heap for the hash calculations. Requiring no garbage collection might set the bar a bit too high.